### DISCUSSION TITLE
Should we commit generated files and how does it affect change detection?

### DISCUSSION THREAD

**Discussion by cwe1ss** [terramate-io/terramate#1108] [Q&A] [answered]
Hi everyone. Thanks for this very interesting product! 

Is it recommended to commit any files generated by Terramate to Git, or should we keep them in a .gitignore and generate them on-demand in CI/CD before we do the plan/apply?

My main concern with committing them is that it might result in orphaned generated files when I e.g. change the file name of a generated file, etc.

Would change detection still work if I keep them in .gitignore?

Thanks in advance to anyone answering!

**Comment by mariux** [comment, marked as answer]
Change detection is highly dependent on the files being checked in in the current version of terramate.

In addition terramate might change generated code slightly in different versions or in worst case and edge case triggers a bug, that could be detected by changes in the generated code easily. Without checking in the generated code won't be available and any issues in generated code is invisible to a reviewer.

We have a feature planned in the backlog that would only generate a checksum that you would still need to check in but that can provide you with slightly better safeguards and supports full change detection as the checksum will change for different versions. In addition the "oudated code safeguard" can trigger accordingly.

We prefix our generated files with e.g. `generated_` so that people can easily skip them in code reviews but if needed debugging an unexpected planned change is easily spotted by any reviewer without the need to locally check out and run `generate`.

Would such a feature just checking in checksum benefit you too?

**Reply by cwe1ss** [reply]
Thank you very much for the good explanation @mariux ! 

I've seen in my tests that Terramate somehow understands which generated files should be deleted, so this seems to address my main concern regarding orphaned files and I'm therefore happy to commit the generated files. We'll probably also use a prefix for the file names - thank you for that suggestion. 

May I ask you to explain how Terramate knows which previously generated files are no longer needed and can therefore be deleted? Does it use Git and compare the previous code?

> Would such a feature just checking in checksum benefit you too?

I don't yet understand the implications enough to say if having just a checksum-file would benefit me somehow.

**Reply by i4ki** [reply]
Hi @cwe1ss 

Sorry for the late reply.

> May I ask you to explain how Terramate knows which previously generated files are no longer needed and can therefore be deleted? Does it use Git and compare the previous code?

Whenever you execute `terramate generate` we scan all `.tm` files and load/eval the `condition` of each `generate_hcl/generate_file` block and then based on the `condition` we know the files that need to be preserved (or created) on disk and the ones that could be deleted. Additionally, for the case of deletion of `generate_hcl` targeted files, we check if the existing file has the Terramate header and we never delete the file if that's not the case. We also use the header to figure out which `generate_hcl` file is orphaned by detecting files that contain the Terramate header but there's no `generate_hcl` block targeting it anymore, then it could be safely deleted.

There are more specific details in the process but this is a high-level overview.
Let me know if it helps.

**Reply by cwe1ss** [reply]
This is really helpful, thank you so much @i4ki !!

**Comment by martinohansen** [comment]
Sorry for hijacking this: is it the intention that the need for checking in the generated files will go away in the future?

**Reply by mariux** [reply]
For Change detection we need to have actual changes in the stacks. We have plans to store a checksum of all generated code instead of the actual files. This will enable us to still detect changes based on a changing checksum but limit the changes in a PR to a single line per stack.

In general we recommend to check in code as it is easier to follow errors and for reviewers to comment on the generated code.

But the answer is yes, we have solution in the backlog to have code-generation on-the-fly and still have change detection available. Sadly there is no ETA. How important would such a feature be for you?
